lib/SILOptimizer/Transforms/DifferentiabilityWitnessDevirtualizer.cpp

```diff
@@ -0,0 +1,71 @@
+//===--- DifferentiabilityWitnessDevirtualizer.cpp ------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Devirtualizes `differentiability_witness_function` instructions into
+// `function_ref` instructions for differentiability witness definitions.
```

Comment:
Some context on why `DifferentiabilityWitnessDevirtualizer` was created as a new transform instead of augmenting `swift::tryDevirtualizeApply`: https://github.com/apple/swift/pull/28480#issuecomment-558467911.

> > Given the similarity to the SIL devirtualizer, maybe it's easier to just add some logic in swift::tryDevirtualizeApply that handles DifferentiabilityWitnessFunction instructions.
> 
> Currently this would be pretty involved because the SIL devirtualizer starts at apply methods and looks for the callee. A lot of the callees in differentiation cases are hidden behind differentiable_function and differentiable_function_extracts.
> 
> The pass as written handles the devirtualization at the reference sites, avoiding this difficulty.